File: android/app/src/main/java/com/livego/app/videocall/CallInvitationService.kt
package com.livego.app.videocall

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Serviço de convite de chamada para Android
 * Gerencia convites via API REST e WebSocket
 */
class CallInvitationService private constructor(private val context: Context) {
    
    companion object {
        private const val TAG = "CallInvitationService"
        private const val PREFS_NAME = "call_invitation_prefs"
        private const val BASE_URL = "https://livego.store"
        
        @Volatile
        private var INSTANCE: CallInvitationService? = null
        
        fun getInstance(context: Context): CallInvitationService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: CallInvitationService(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val gson = Gson()
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    
    // Estado da chamada atual
    private val _currentCall = MutableStateFlow<CallInvitation?>(null)
    val currentCall: StateFlow<CallInvitation?> = _currentCall.asStateFlow()
    
    // Estado da conexão
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()
    
    data class CallInvitation(
        val id: String,
        val hostId: String,
        val hostName: String,
        val guestId: String?,
        val guestName: String?,
        val roomId: String,
        val streamId: String,
        val streamTitle: String?,
        val token: String?,
        val wsUrl: String?
    )
    
    data class InviteRequest(
        val guestId: String,
        val guestName: String,
        val streamId: String
    )
    
    data class RespondRequest(
        val invitationId: String,
        val response: String // "accept" ou "decline"
    )
    
    data class ApiResponse<T>(
        val success: Boolean,
        val error: String?,
        val data: T? = null
    )
    
    /**
     * Convidar usuário para entrar na live
     */
    suspend fun inviteGuest(
        userToken: String,
        guestId: String,
        guestName: String,
        streamId: String
    ): Result<CallInvitation> {
        return try {
            val request = InviteRequest(guestId, guestName, streamId)
            val json = gson.toJson(request)
            val body = json.toRequestBody("application/json".toMediaType())
            
            val httpRequest = Request.Builder()
                .url("$BASE_URL/api/call-invitation/invite")
                .addHeader("Authorization", "Bearer $userToken")
                .addHeader("Content-Type", "application/json")
                .post(body)
                .build()
            
            client.newCall(httpRequest).execute().use { response ->
                val responseBody = response.body?.string()
                if (response.isSuccessful && responseBody != null) {
                    val apiResponse = gson.fromJson(responseBody, object : TypeToken<ApiResponse<Map<String, String>>>() {}.type)
                    if (apiResponse.success) {
                        val invitationId = apiResponse.data?.get("invitationId") ?: ""
                        val invitation = CallInvitation(
                            id = invitationId,
                            hostId = getCurrentUserId(),
                            hostName = getCurrentUserName(),
                            guestId = guestId,
                            guestName = guestName,
                            roomId = "", // Será preenchido quando aceito
                            streamId = streamId,
                            streamTitle = null,
                            token = null,
                            wsUrl = null
                        )
                        _currentCall.value = invitation
                        Result.success(invitation)
                    } else {
                        Result.failure(Exception(apiResponse.error ?: "Erro desconhecido"))
                    }
                } else {
                    Result.failure(Exception("Erro na requisição: ${response.code}"))
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Erro ao convidar usuário", e)
            Result.failure(e)
        }
    }
    
    /**
     * Responder a um convite
     */
    suspend fun respondToInvitation(
        userToken: String,
        invitationId: String,
        response: String
    ): Result<String> {
        return try {
            val request = RespondRequest(invitationId, response)
            val json = gson.toJson(request)
            val body = json.toRequestBody("application/json".toMediaType())
            
            val httpRequest = Request.Builder()
                .url("$BASE_URL/api/call-invitation/respond")
                .addHeader("Authorization", "Bearer $userToken")
                .addHeader("Content-Type", "application/json")
                .post(body)
                .build()
            
            client.newCall(httpRequest).execute().use { response ->
                val responseBody = response.body?.string()
                if (response.isSuccessful && responseBody != null) {
                    val apiResponse = gson.fromJson(responseBody, object : TypeToken<ApiResponse<Map<String, String>>>() {}.type)
                    if (apiResponse.success) {
                        Result.success(apiResponse.data?.get("status") ?: response)
                    } else {
                        Result.failure(Exception(apiResponse.error ?: "Erro desconhecido"))
                    }
                } else {
                    Result.failure(Exception("Erro na requisição: ${response.code}"))
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Erro ao responder convite", e)
            Result.failure(e)
        }
    }
    
    /**
     * Encerrar chamada
     */
    suspend fun endCall(
        userToken: String,
        invitationId: String
    ): Result<Unit> {
        return try {
            val json = JSONObject().apply {
                put("invitationId", invitationId)
            }.toString()
            val body = json.toRequestBody("application/json".toMediaType())
            
            val httpRequest = Request.Builder()
                .url("$BASE_URL/api/call-invitation/end")
                .addHeader("Authorization", "Bearer $userToken")
                .addHeader("Content-Type", "application/json")
                .post(body)
                .build()
            
            client.newCall(httpRequest).execute().use { response ->
                val responseBody = response.body?.string()
                if (response.isSuccessful && responseBody != null) {
                    val apiResponse = gson.fromJson(responseBody, object : TypeToken<ApiResponse<Unit>>() {}.type)
                    if (apiResponse.success) {
                        _currentCall.value = null
                        Result.success(Unit)
                    } else {
                        Result.failure(Exception(apiResponse.error ?: "Erro desconhecido"))
                    }
                } else {
                    Result.failure(Exception("Erro na requisição: ${response.code}"))
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Erro ao encerrar chamada", e)
            Result.failure(e)
        }
    }
    
    /**
     * Obter token do usuário atual
     */
    private fun getCurrentUserToken(): String? {
        return prefs.getString("user_token", null)
    }
    
    /**
     * Obter ID do usuário atual
     */
    private fun getCurrentUserId(): String {
        return prefs.getString("user_id", "") ?: ""
    }
    
    /**
     * Obter nome do usuário atual
     */
    private fun getCurrentUserName(): String {
        return prefs.getString("user_name", "") ?: ""
    }
    
    /**
     * Salvar chamada atual
     */
    fun saveCurrentCall(call: CallInvitation) {
        val json = gson.toJson(call)
        prefs.edit()
            .putString("current_call", json)
            .apply()
        _currentCall.value = call
    }
    
    /**
     * Obter chamada atual salva
     */
    fun getCurrentCall(): CallInvitation? {
        val json = prefs.getString("current_call", null)
        return if (json != null) {
            gson.fromJson(json, CallInvitation::class.java)
        } else null
    }
    
    /**
     * Limpar chamada atual
     */
    fun clearCurrentCall() {
        prefs.edit()
            .remove("current_call")
            .apply()
        _currentCall.value = null
    }
    
    /**
     * Verificar se está em chamada
     */
    fun isInCall(): Boolean {
        return _currentCall.value != null
    }
    
    /**
     * Atualizar estado da conexão
     */
    fun setConnected(connected: Boolean) {
        _isConnected.value = connected
    }
    
    /**
     * Limpar recursos
     */
    fun cleanup() {
        clearCurrentCall()
        _isConnected.value = false
    }
}

File: src/services/CallInvitationService.ts
import { io, Socket } from 'socket.io-client';
import { getUserIdFromToken } from '../utils/auth';

export interface CallInvitation {
  id: string;
  hostId: string;
  hostName: string;
  guestId?: string;
  guestName?: string;
  roomId: string;
  streamId: string;
  streamTitle?: string;
  token?: string;
  wsUrl?: string;
}

export interface CallInvitationEvent {
  type: 'invitation_received' | 'invitation_sent' | 'invitation_accepted' | 'invitation_declined' | 'call_joined' | 'call_ended';
  invitation: CallInvitation;
}

class CallInvitationService {
  private socket: Socket | null = null;
  private currentCall: CallInvitation | null = null;
  private listeners: Map<string, Function[]> = new Map();

  constructor() {
    this.initializeSocket();
  }

  private initializeSocket() {
    const token = localStorage.getItem('token');
    if (!token) return;

    this.socket = io('https://livego.store', {
      auth: { token },
      transports: ['websocket', 'polling']
    });

    this.setupEventListeners();
  }

  private setupEventListeners() {
    if (!this.socket) return;

    // Entrar na sala do usuário para receber convites
    const userId = getUserIdFromToken();
    if (userId) {
      this.socket.emit('join_user_room', userId);
    }

    // Listener para convites de chamada
    this.socket.on('call_invitation', (event: CallInvitationEvent) => {
      this.handleCallInvitation(event);
    });

    // Listener para conexão
    this.socket.on('connect', () => {
      console.log('📞 Conectado ao serviço de convites');
      const userId = getUserIdFromToken();
      if (userId) {
        this.socket?.emit('join_user_room', userId);
      }
    });

    // Listener para desconexão
    this.socket.on('disconnect', () => {
      console.log('📞 Desconectado do serviço de convites');
    });
  }

  private handleCallInvitation(event: CallInvitationEvent) {
    console.log('📞 Evento de chamada:', event);

    // Atualizar estado atual da chamada
    switch (event.type) {
      case 'invitation_received':
        this.currentCall = event.invitation;
        break;
      case 'invitation_accepted':
        if (event.invitation.token) {
          this.currentCall = event.invitation;
          this.connectToLiveKit(event.invitation);
        }
        break;
      case 'call_ended':
        this.currentCall = null;
        this.disconnectFromLiveKit();
        break;
    }

    // Notificar listeners
    this.notifyListeners('callInvitation', event);
  }

  // Convidar usuário para entrar na live
  async inviteGuest(guestId: string, guestName: string, streamId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const token = localStorage.getItem('token');
      if (!token) {
        return { success: false, error: 'Token não encontrado' };
      }

      const response = await fetch('https://livego.store/api/call-invitation/invite', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify({
          guestId,
          guestName,
          streamId
        })
      });

      const data = await response.json();
      return data;
    } catch (error: any) {
      console.error('Erro ao convidar usuário:', error);
      return { success: false, error: error.message };
    }
  }

  // Responder a um convite
  async respondToInvitation(invitationId: string, response: 'accept' | 'decline'): Promise<{ success: boolean; error?: string }> {
    try {
      const token = localStorage.getItem('token');
      if (!token) {
        return { success: false, error: 'Token não encontrado' };
      }

      const apiResponse = await fetch('https://livego.store/api/call-invitation/respond', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify({
          invitationId,
          response
        })
      });

      const data = await apiResponse.json();
      return data;
    } catch (error: any) {
      console.error('Erro ao responder convite:', error);
      return { success: false, error: error.message };
    }
  }

  // Encerrar chamada
  async endCall(invitationId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const token = localStorage.getItem('token');
      if (!token) {
        return { success: false, error: 'Token não encontrado' };
      }

      const response = await fetch('https://livego.store/api/call-invitation/end', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify({
          invitationId
        })
      });

      const data = await response.json();
      return data;
    } catch (error: any) {
      console.error('Erro ao encerrar chamada:', error);
      return { success: false, error: error.message };
    }
  }

  // Conectar ao LiveKit quando o convite é aceito
  private async connectToLiveKit(invitation: CallInvitation) {
    try {
      if (!invitation.token || !invitation.wsUrl) {
        console.error('Token ou WS URL não encontrados no convite');
        return;
      }

      // Importar e usar LiveKit
      const { Room } = await import('livekit-client');
      
      const room = new Room();
      
      await room.connect(invitation.wsUrl, invitation.token);
      
      console.log('📞 Conectado à sala LiveKit:', invitation.roomId);
      
      // Notificar que está conectado
      this.notifyListeners('connected', { room, invitation });
      
    } catch (error: any) {
      console.error('Erro ao conectar ao LiveKit:', error);
      this.notifyListeners('error', { error: error.message });
    }
  }

  // Desconectar do LiveKit
  private disconnectFromLiveKit() {
    // Implementar desconexão do LiveKit
    console.log('📞 Desconectado do LiveKit');
    this.notifyListeners('disconnected', {});
  }

  // Adicionar listener para eventos
  addListener(event: string, callback: Function) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
    }
    this.listeners.get(event)!.push(callback);
  }

  // Remover listener
  removeListener(event: string, callback: Function) {
    const callbacks = this.listeners.get(event);
    if (callbacks) {
      const index = callbacks.indexOf(callback);
      if (index > -1) {
        callbacks.splice(index, 1);
      }
    }
  }

  // Notificar todos os listeners de um evento
  private notifyListeners(event: string, data: any) {
    const callbacks = this.listeners.get(event);
    if (callbacks) {
      callbacks.forEach(callback => callback(data));
    }
  }

  // Obter chamada atual
  getCurrentCall(): CallInvitation | null {
    return this.currentCall;
  }

  // Verificar se está em chamada
  isInCall(): boolean {
    return this.currentCall !== null;
  }

  // Limpar recursos
  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
    this.currentCall = null;
    this.listeners.clear();
  }
}

// Singleton
export const callInvitationService = new CallInvitationService();
export default callInvitationService;
